executor/aggregate.go

```diff
@@ -45,39 +143,494 @@ type HashAggExec struct {
+			inputCh:           e.partialInputChs[i],
+			outputChs:         e.partialOutputChs,
+			giveBackCh:        e.inputCh,
+			globalOutputCh:    e.finalOutputCh,
+			aggCtxsMap:        make(aggCtxsMapper, 0),
+			groupByItems:      e.GroupByItems,
+			chk:               e.children[0].newChunk(),
+		}
+
+		e.partialWorkers[i] = w
+		e.inputCh <- &HashAggInput{
+			chk:        e.children[0].newChunk(),
+			giveBackCh: w.inputCh,
+		}
+	}
+
+	// Init final workers.
+	for i := 0; i < finalConcurrency; i++ {
+		e.finalWorkers[i] = HashAggFinalWorker{
+			baseHashAggWorker:   newBaseHashAggWorker(e.finishCh, e.newFinalAggFuncs(), e.maxChunkSize),
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupSet:            mvmap.NewMVMap(),
+			groupVals:           make([][]byte, 0, 8),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+func (w *HashAggPartialWorker) getChildInput() bool {
+	select {
+	case <-w.finishCh:
+		return false
+	case chk, ok := <-w.inputCh:
+		if !ok {
+			return false
+		}
+		w.chk.SwapColumns(chk)
+		w.giveBackCh <- &HashAggInput{
+			chk:        chk,
+			giveBackCh: w.inputCh,
+		}
+	}
+	return true
+}
+
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup, finalConcurrency int) {
+	needShuffle, sc := false, ctx.GetSessionVars().StmtCtx
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc, finalConcurrency)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		if !w.getChildInput() {
+			return
+		}
+		if err := w.updatePartialResult(ctx, sc, w.chk, len(w.aggCtxsMap)); err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		// The intermData can be promised to be not empty if reaching here,
+		// so we set needShuffle to be true.
+		needShuffle = true
+	}
+}
+
+func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (err error) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, w.aggCtxsMap)
+		for i, af := range w.aggFuncs {
+			if af.Update(aggEvalCtxs[i], sc, row) != nil {
+				return errors.Trace(err)
```

Comment:
Why ignore the err of af.Update? Should it be `if err = af.Update(aggEvalCtxs[i], sc, row); err != nil`?